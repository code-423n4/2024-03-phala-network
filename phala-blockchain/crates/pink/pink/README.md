<h1 align="center">Phala's ink! - Writing Enhanced Smart Contracts</h1>

Pink! is a smart contract language extending Parity's `ink!`. It extends the basic functionality with additional features, tailored to interact efficiently with Phala's Phat Contract runtime.

## Getting Started

Unaltered `ink!` contracts are fully compatible and executable on the Phala's Phat Contract platform. To learn how to start writing contracts with `ink!`, follow the Parity's [ink! documentation](https://paritytech.github.io/ink-docs/).

To get started with Pink!, add the following dependency to your `Cargo.toml`:

```toml
[dependencies]
ink = { version = "4", default-features = false }
pink = { version = "0.4", default-features = false }

[features]
std = [
    "ink/std",
    "pink/std",
]
```

Then, you can use the `http_get!` macro to make a GET request to a remote server:

```ignore
#[ink::message]
fn http_get_example(&self) {
    let response = pink::http_get!("https://httpbin.org/get");
    assert_eq!(response.status_code, 200);
}
```

## Phat Contract-Specific Features

The Pink! crate is designed to empower you, enabling you to leverage the unique features of the Phat Contract, such as making HTTP requests as demonstrated in our examples. This crate supplies the crucial types and functions needed to seamlessly interact with the Phat Contract runtime.

There are three kind of APIs to communication with the runtime:

-   Emitting Events:
    These APIs are primarily used in situations where the operation could lead to side effects that need to be deterministically recorded and may be rolled back during the execution of the contract call. For additional information on Emitting Events APIs, please refer to the [PinkEvent documentation](crate::PinkEvent).

-   Chain Extension:
    These APIs are predominantly used for read-only operations or operations that aren't expected to create deterministic side effects. For an in-depth understanding of Chain Extension APIs, please refer to the [PinkExt documentation](crate::chain_extension::PinkExtBackend).

-   System contract:
    There is a special contract called the System contract in each cluster. The system contract is instantiated when the cluster is created. Either ink contracts or external accounts can call the system contract to perform certain operations. For more information on the System contract, please refer to the [System documentation](crate::system::SystemForDoc).

For practical implementation examples, explore our [Phat Contract examples](https://github.com/Phala-Network/phat-contract-examples) repository.

## Using JavaScript with Phat Contract

Phat Contract supports JavaScript through the [phat-quickjs](https://github.com/Phala-Network/phat-quickjs) contract.

There are two ways to use JavaScript in your contract:

-   You can deploy your phat-quickjs contract instance through a standard deployment process.

-   However, for a more convenient approach, most public clusters should already have a public driver quickjs contract deployed. You can obtain the contract code_hash with `System::get_driver("JsDelegate")`.

-   For the simplest integration, consider using the [`phat_js`](https://docs.rs/phat_js/) crate. It provides an `eval` function that lets you evaluate JavaScript code snippets directly.
    For example:
    ```ignore
    #[ink::message]
    fn eval_js_example(&self) {
        let result = phat_js::eval("'Hello,' + 'World'", &[]);
        assert_eq!(result, phat_js::Output::String("Hello,World".to_string()));
    }
    ```
