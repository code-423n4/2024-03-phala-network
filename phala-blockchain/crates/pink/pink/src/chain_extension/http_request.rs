use alloc::string::String;
use alloc::vec::Vec;
use num_enum::{IntoPrimitive, TryFromPrimitive};

use super::ErrorCode;
#[derive(scale::Encode, scale::Decode)]
#[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
pub struct HttpRequest {
    pub url: String,
    pub method: String,
    pub headers: Vec<(String, String)>,
    pub body: Vec<u8>,
}

impl HttpRequest {
    /// Create a new http request.
    pub fn new(
        url: impl Into<String>,
        method: impl Into<String>,
        headers: Vec<(String, String)>,
        body: Vec<u8>,
    ) -> Self {
        Self {
            url: url.into(),
            method: method.into(),
            headers,
            body,
        }
    }
}

#[derive(scale::Encode, scale::Decode)]
#[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
pub struct HttpResponse {
    pub status_code: u16,
    pub reason_phrase: String,
    pub headers: Vec<(String, String)>,
    pub body: Vec<u8>,
}

#[derive(
    scale::Encode, scale::Decode, TryFromPrimitive, IntoPrimitive, Clone, Copy, Debug, PartialEq, Eq,
)]
#[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
#[repr(u32)]
pub enum HttpRequestError {
    InvalidUrl,
    InvalidMethod,
    InvalidHeaderName,
    InvalidHeaderValue,
    FailedToCreateClient,
    Timeout,
    NotAllowed,
    TooManyRequests,
    NetworkError,
    ResponseTooLarge,
}

impl super::sealed::Sealed for HttpRequestError {}
impl super::CodableError for HttpRequestError {
    fn encode(&self) -> u32 {
        let code: u32 = (*self).into();
        code + 1
    }

    fn decode(code: u32) -> Option<Self> {
        if code == 0 {
            return None;
        }
        core::convert::TryFrom::try_from(code - 1).ok()
    }
}

impl From<ErrorCode> for HttpRequestError {
    fn from(value: ErrorCode) -> Self {
        match <Self as super::CodableError>::decode(value.0) {
            None => crate::panic!("chain extension: invalid HttpRequestError"),
            Some(err) => err,
        }
    }
}

impl From<scale::Error> for HttpRequestError {
    fn from(_value: scale::Error) -> Self {
        crate::panic!("chain_ext: failed to decocde HttpRequestError")
    }
}

impl HttpRequestError {
    pub fn display(&self) -> &'static str {
        match self {
            Self::InvalidUrl => "Invalid URL",
            Self::InvalidMethod => "Invalid method",
            Self::InvalidHeaderName => "Invalid header name",
            Self::InvalidHeaderValue => "Invalid header value",
            Self::FailedToCreateClient => "Failed to create client",
            Self::Timeout => "Timeout",
            Self::NotAllowed => "Not allowed",
            Self::TooManyRequests => "Too many requests",
            Self::NetworkError => "Network error",
            Self::ResponseTooLarge => "Response too large",
        }
    }
}

impl HttpResponse {
    pub fn ok(body: Vec<u8>) -> Self {
        Self {
            status_code: 200,
            reason_phrase: "OK".into(),
            headers: Default::default(),
            body,
        }
    }

    pub fn not_found() -> Self {
        Self {
            status_code: 404,
            reason_phrase: "Not Found".into(),
            headers: Default::default(),
            body: Default::default(),
        }
    }
}

#[macro_export]
macro_rules! http_req {
    ($method: expr, $url: expr, $data: expr, $headers: expr) => {{
        use $crate::chain_extension::HttpRequest;
        let headers = $headers;
        let body = $data;
        let request = HttpRequest {
            url: $url.into(),
            method: $method.into(),
            headers,
            body,
        };
        $crate::ext().http_request(request)
    }};
    ($method: expr, $url: expr, $data: expr) => {{
        $crate::http_req!($method, $url, $data, Default::default())
    }};
}

/// Make a simple HTTP GET request
///
/// # Arguments
/// - url: The URL to GET
/// - headers: The headers to send with the request
///
/// # Examples
///
/// ```ignore
/// use pink::http_get;
/// let response = http_get!("https://example.com/");
/// assert_eq!(response.status_code, 200);
/// ```
///
/// ```ignore
/// use pink::http_get;
/// let headers = vec![("X-Foo".into(), "Bar".into())];
/// let response = http_get!("https://example.com/", headers);
/// assert_eq!(response.status_code, 200);
/// ```
#[macro_export]
macro_rules! http_get {
    ($url: expr, $headers: expr) => {{
        $crate::http_req!("GET", $url, Default::default(), $headers)
    }};
    ($url: expr) => {{
        $crate::http_get!($url, Default::default())
    }};
}

/// Make a simple HTTP POST request
///
/// # Arguments
/// - url: The URL to POST
/// - data: The payload to POST
/// - headers: The headers to send with the request
///
/// # Examples
///
/// ```ignore
/// use pink::http_post;
/// let response = http_post!("https://example.com/", b"Hello, world!");
/// assert_eq!(response.status_code, 200);
/// ```
///
/// ```ignore
/// use pink::http_post;
/// let headers = vec![("X-Foo".into(), "Bar".into())];
/// let response = http_post!("https://example.com/", b"Hello, world!", headers);
/// assert_eq!(response.status_code, 200);
/// ```
#[macro_export]
macro_rules! http_post {
    ($url: expr, $data: expr, $headers: expr) => {{
        $crate::http_req!("POST", $url, $data.into(), $headers)
    }};
    ($url: expr, $data: expr) => {{
        $crate::http_post!($url, $data, Default::default())
    }};
}

/// Make a simple HTTP PUT request
///
/// # Arguments
/// - url: The destination URL
/// - data: The payload to PUT
/// - headers: The headers to send with the request
///
/// # Examples
///
/// ```ignore
/// use pink::http_put;
/// let response = http_put!("https://example.com/", b"Hello, world!");
/// assert_eq!(response.status_code, 200);
/// ```
#[macro_export]
macro_rules! http_put {
    ($url: expr, $data: expr, $headers: expr) => {{
        $crate::http_req!("PUT", $url, $data.into(), $headers)
    }};
    ($url: expr, $data: expr) => {{
        $crate::http_put!($url, $data, Default::default())
    }};
}
