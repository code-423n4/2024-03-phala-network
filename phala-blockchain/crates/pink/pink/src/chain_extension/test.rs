use scale::Decode;

pub struct MockExtensionFn<F, I, const FID: u32> {
    call: F,
    _p: std::marker::PhantomData<(I,)>,
}

impl<F, In, const FID: u32> ink::env::test::ChainExtension for MockExtensionFn<F, In, FID>
where
    In: Decode,
    F: FnMut(In) -> (u32, Vec<u8>),
{
    fn func_id(&self) -> u32 {
        FID
    }

    fn call(&mut self, input: &[u8], output: &mut Vec<u8>) -> u32 {
        let input: Vec<u8> = Decode::decode(&mut &input[..]).expect("mock decode input failed");
        let input = In::decode(&mut &input[..]).expect("mock decode input failed");
        let (status, data) = (self.call)(input);
        output.extend(&data);
        status
    }
}

impl<F, In, const FID: u32> MockExtensionFn<F, In, FID>
where
    In: Decode,
    F: FnMut(In) -> (u32, Vec<u8>),
{
    pub fn new(call: F) -> Self {
        Self {
            call,
            _p: Default::default(),
        }
    }
}
