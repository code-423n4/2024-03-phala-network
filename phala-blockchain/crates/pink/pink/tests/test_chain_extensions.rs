use ink::env::chain_extension::FromStatusCode;
use pink::{
    chain_extension::{
        CodableError, ErrorCode, HttpRequest, HttpRequestError, HttpResponse, JsCode, JsValue,
        StorageQuotaExceeded,
    },
    HookPoint, PinkEvent, ResultExt, SidevmOperation,
};
use pink_chain_extension::{local_cache::apply_quotas, mock_ext::mock_all_ext};
use rusty_fork::rusty_fork_test;
use scale::{Decode, Encode};
use std::convert::TryInto;

#[test]
fn getrandom_works() {
    mock_all_ext();

    let bytes = pink::ext().getrandom(3);
    assert_eq!(bytes.len(), 3);
    assert!(bytes != [0; 3]);
}

#[test]
fn test_signing() {
    use pink::chain_extension::signing as sig;
    use pink::chain_extension::SigType;

    mock_all_ext();

    let privkey = sig::derive_sr25519_key(b"a spoon of salt");
    let pubkey = sig::get_public_key(&privkey, SigType::Sr25519);
    let message = b"hello world";
    let signature = sig::sign(message, &privkey, SigType::Sr25519);
    let pass = sig::verify(message, &pubkey, &signature, SigType::Sr25519);
    assert!(pass);
    let pass = sig::verify(b"Fake", &pubkey, &signature, SigType::Sr25519);
    assert!(!pass);
}

#[test]
fn test_ecdsa_signing() {
    use pink::chain_extension::signing as sig;
    use pink::chain_extension::SigType;

    mock_all_ext();

    let privkey = sig::derive_sr25519_key(b"salt");
    let privkey = &privkey[..32];
    let pubkey: pink::EcdsaPublicKey = sig::get_public_key(privkey, SigType::Ecdsa)
        .try_into()
        .unwrap();
    let message = [1u8; 32];
    let signature = sig::ecdsa_sign_prehashed(privkey, message);
    let pass = sig::ecdsa_verify_prehashed(signature, message, pubkey);
    let fake_message = [2u8; 32];
    assert!(pass);
    let pass = sig::ecdsa_verify_prehashed(signature, fake_message, pubkey);
    assert!(!pass);
}

#[test]
fn local_cache_works_thread0() {
    mock_all_ext();

    assert!(pink::ext().cache_set(b"foo", b"bar-0").is_ok());
    let value = pink::ext().cache_get(b"foo");
    assert_eq!(value, Some(b"bar-0".to_vec()));
    pink::ext().cache_remove(b"foo");
    assert_eq!(pink::ext().cache_get(b"foo"), None);
}

#[test]
fn local_cache_works_thread1() {
    mock_all_ext();

    assert!(pink::ext().cache_set(b"foo", b"bar-1").is_ok());
    let value = pink::ext().cache_get(b"foo");
    assert_eq!(value, Some(b"bar-1".to_vec()));
    pink::ext().cache_set_expiration(b"foo", 0);
    assert_eq!(pink::ext().cache_get(b"foo"), None);
}

#[test]
fn local_cache_works_thread2() {
    mock_all_ext();

    assert!(pink::ext().cache_set(b"foo", b"bar-2").is_ok());
    let value = pink::ext().cache_get(b"foo");
    assert_eq!(value, Some(b"bar-2".to_vec()));
}

#[test]
fn test_systime() {
    use std::time::SystemTime;
    mock_all_ext();

    let ms = pink::ext().untrusted_millis_since_unix_epoch();
    let now = SystemTime::now()
        .duration_since(SystemTime::UNIX_EPOCH)
        .unwrap()
        .as_millis();
    assert!((now as i128 - ms as i128).abs() < 100);
}

#[cfg(coverage)]
#[test]
fn test_http_req() {
    mock_all_ext();

    let resp = pink::http_get!(
        "https://httpbin.org/get",
        vec![("X-Foo".into(), "Bar".into())]
    );
    assert_eq!(resp.status_code, 200);

    let resp = pink::http_post!(
        "https://httpbin.org/post",
        *b"hello world",
        vec![("X-Foo".into(), "Bar".into())]
    );
    assert_eq!(resp.status_code, 200);
}

#[test]
fn test_http_invalid_url() {
    mock_all_ext();
    let request = HttpRequest::new(
        "sip://httpbin.org/get",
        "GET",
        vec![("X-Foo".into(), "Bar".into())],
        b"".to_vec(),
    );
    let resp = pink::ext().http_request(request);
    assert_eq!(resp.status_code, 523);
}

#[test]
fn test_chores() {
    mock_all_ext();

    assert!(HttpResponse::not_found().encode() != HttpResponse::ok(vec![]).encode());
    use HttpRequestError::*;

    for err in [
        InvalidUrl,
        InvalidMethod,
        InvalidHeaderName,
        InvalidHeaderValue,
        FailedToCreateClient,
        Timeout,
        NotAllowed,
        TooManyRequests,
        NetworkError,
        ResponseTooLarge,
    ] {
        let encoded = CodableError::encode(&err);
        let decoded = HttpRequestError::from(ErrorCode::from_status_code(encoded).unwrap_err());
        assert_eq!(err, decoded);
        assert!(!decoded.display().is_empty());
        assert!(!format!("{:?}", decoded).is_empty());

        let scl = Encode::encode(&decoded.clone());
        let scl_decoded: HttpRequestError = Decode::decode(&mut &scl[..]).unwrap();
        assert_eq!(err, scl_decoded);
    }

    assert!(!pink::ext().is_in_transaction());
    assert_eq!(pink::ext().worker_pubkey(), [0u8; 32]);
    assert!(!pink::ext().code_exists([1; 32], false));
    assert_eq!(pink::ext().runtime_version(), (1, 0));
    assert_eq!(pink::ext().current_event_chain_head().0, 0);
    let rv = pink::ext().js_eval(vec![], vec![]);
    assert_eq!(rv, JsValue::Exception("No Js Runtime".to_string()));
}

#[test]
#[should_panic]
fn test_get_system_contract_id() {
    mock_all_ext();
    let _ = pink::ext().system_contract_id();
}

#[test]
fn dump_type_info() {
    use type_info_stringify::type_info_stringify;

    #[derive(scale_info::TypeInfo)]
    struct Root {
        _http_request: HttpRequest,
        _http_response: HttpResponse,
        _http_error: HttpRequestError,
        _quota_error: StorageQuotaExceeded,
        _error_code: ErrorCode,
        _js_code: JsCode,
        _js_value: JsValue,
        _pink_event: PinkEvent,
        _system_err: pink::system::Error,
        _driver_err: pink::system::DriverError,
        _code_type: pink::system::CodeType,
    }

    insta::assert_snapshot!(type_info_stringify::<Root>())
}

#[test]
fn test_codecs() {
    use scale::{Decode, Encode};
    let encoded = Encode::encode(&StorageQuotaExceeded);
    assert!(encoded.is_empty());
    let _decoded = <StorageQuotaExceeded as Decode>::decode(&mut &encoded[..]).unwrap();
    assert_eq!(
        format!("{:?}", StorageQuotaExceeded),
        "StorageQuotaExceeded"
    );
    assert!(<StorageQuotaExceeded as CodableError>::decode(0).is_none());
}

rusty_fork_test! {
    #[test]
    fn cache_quotas_works() {
        tracing_subscriber::fmt::init();
        mock_all_ext();
        apply_quotas(vec![(&b""[..], 10)]);
        assert!(pink::ext().cache_set(b"foo", b"bar-0").is_ok());
        assert!(pink::ext().cache_get(b"foo").is_some());
        assert!(pink::ext().cache_set(b"bar", b"baaaaaaaaaaaaaaaar").is_err());
        assert!(pink::ext().cache_get(b"bar").is_none());
    }

    #[test]
    fn trigger_logs() {
        tracing_subscriber::fmt::init();
        mock_all_ext();

        pink::error!("pink error");
        pink::warn!("pink warn");
        pink::info!("pink info");
        pink::debug!("pink debug");
        pink::trace!("pink trace");

        _ = Result::<(), &str>::Ok(()).log_err("Hello");
        _ = Result::<(), &str>::Err("Error message").log_err("Hello");
        _ = Result::<(), &str>::Ok(())
            .log_err_with_level(pink::logger::Level::Info, "Hello");
        _ = Result::<(), &str>::Err("Error message")
            .log_err_with_level(pink::logger::Level::Info, "Hello");
    }
}

#[test]
fn cov_emit_events() {
    mock_all_ext();

    pink::deploy_sidevm_to([0u8; 32].into(), [0; 32]);
    pink::force_stop_sidevm();
    pink::push_sidevm_message(b"foo".to_vec());
    pink::set_contract_weight([0u8; 32].into(), 0);
    pink::set_hook(HookPoint::OnBlockEnd, [0u8; 32].into(), 0, 0);
    pink::set_js_runtime([0u8; 32]);
    pink::set_log_handler([0u8; 32].into());
    pink::stop_sidevm_at([0u8; 32].into());
    pink::upgrade_runtime((0, 1));

    let result = pink::query_local_sidevm([0u8; 32].into(), vec![]);
    assert!(result.is_err());
}

#[test]
fn cov_pink_events_display() {
    fn test_all(events: &[(bool, bool, &str, PinkEvent)]) {
        for (is_private, allowed_in_query, name, event) in events {
            assert_eq!(event.name(), *name);
            assert_eq!(event.is_private(), *is_private);
            assert_eq!(event.allowed_in_query(), *allowed_in_query);
        }
    }

    test_all(&[
        // Test for PinkEvent::SetHook variant
        (
            false,
            false,
            "SetHook",
            PinkEvent::SetHook {
                hook: HookPoint::OnBlockEnd,
                contract: [0u8; 32].into(),
                selector: 0,
                gas_limit: 0,
            },
        ),
        // Test for PinkEvent::DeploySidevmTo variant
        (
            false,
            true,
            "DeploySidevmTo",
            PinkEvent::DeploySidevmTo {
                contract: [0u8; 32].into(),
                code_hash: [0u8; 32],
            },
        ),
        // Test for PinkEvent::SidevmMessage variant
        (
            true,
            true,
            "SidevmMessage",
            PinkEvent::SidevmMessage(vec![]),
        ),
        // Test for PinkEvent::CacheOp variant
        (
            true,
            true,
            "CacheOp",
            PinkEvent::CacheOp(pink::CacheOp::Set {
                key: vec![],
                value: vec![],
            }),
        ),
        // Test for PinkEvent::StopSidevm variant
        (false, true, "StopSidevm", PinkEvent::StopSidevm),
        // Test for PinkEvent::ForceStopSidevm variant
        (
            false,
            true,
            "ForceStopSidevm",
            PinkEvent::ForceStopSidevm {
                contract: [0u8; 32].into(),
            },
        ),
        // Test for PinkEvent::SetLogHandler variant
        (
            false,
            false,
            "SetLogHandler",
            PinkEvent::SetLogHandler([0u8; 32].into()),
        ),
        // Test for PinkEvent::SetContractWeight variant
        (
            false,
            false,
            "SetContractWeight",
            PinkEvent::SetContractWeight {
                contract: [0u8; 32].into(),
                weight: 123,
            },
        ),
        // Test for PinkEvent::UpgradeRuntimeTo variant
        (
            false,
            false,
            "UpgradeRuntimeTo",
            PinkEvent::UpgradeRuntimeTo { version: (1, 0) },
        ),
        // Test for PinkEvent::SidevmOperation - Start variant
        (
            false,
            true,
            "SidevmOperation",
            PinkEvent::SidevmOperation(SidevmOperation::Start {
                contract: [0u8; 32].into(),
                code_hash: [0u8; 32],
                workers: pink::Workers::List(vec![]),
                config: pink::SidevmConfig::default(),
            }),
        ),
        // Test for PinkEvent::SetJsRuntime variant
        (
            false,
            false,
            "SetJsRuntime",
            PinkEvent::SetJsRuntime([0u8; 32]),
        ),
    ])
}

#[test]
fn cov_env() {
    ink::env::test::set_block_timestamp::<pink::PinkEnvironment>(1);
    assert_eq!(pink::env().block_timestamp(), 1);
}

#[test]
fn cov_vrf() {
    mock_all_ext();

    let vrf = pink::vrf(&[]);
    assert_eq!(vrf.len(), 64);
}

#[test]
#[should_panic]
fn cov_start_sidevm() {
    mock_all_ext();

    let result = pink::start_sidevm([0u8; 32]);
    assert!(result.is_err());
}

#[test]
fn cov_system_error() {
    use pink::system::{CodeType, DriverError, Error};

    assert!(matches!(
        DriverError::from(Error::CodeNotFound),
        DriverError::SystemError(_)
    ));

    assert!(CodeType::Ink.is_ink());
    assert!(!CodeType::Ink.is_sidevm());
    assert!(CodeType::Sidevm.is_sidevm());
    assert!(!CodeType::Sidevm.is_ink());
}
