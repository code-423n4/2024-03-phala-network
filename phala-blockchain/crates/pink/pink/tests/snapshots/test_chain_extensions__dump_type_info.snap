---
source: crates/pink/pink/tests/test_chain_extensions.rs
assertion_line: 203
expression: "type_info_stringify::<Root>()"
---
pink::chain_extension::http_request::HttpRequest = struct {
    url: String,
    method: String,
    headers: Vec<(String, String)>,
    body: Vec<u8>,
}
pink::chain_extension::http_request::HttpResponse = struct {
    status_code: u16,
    reason_phrase: String,
    headers: Vec<(String, String)>,
    body: Vec<u8>,
}
pink::chain_extension::http_request::HttpRequestError = enum {
    [0]InvalidUrl,
    [1]InvalidMethod,
    [2]InvalidHeaderName,
    [3]InvalidHeaderValue,
    [4]FailedToCreateClient,
    [5]Timeout,
    [6]NotAllowed,
    [7]TooManyRequests,
    [8]NetworkError,
    [9]ResponseTooLarge,
}
pink::chain_extension::StorageQuotaExceeded = struct {
}
pink::chain_extension::ErrorCode = struct {
    : u32,
}
pink_types::js::JsCode = enum {
    [0]Source(String)
    [1]Bytecode(Vec<u8>)
}
pink_types::js::JsValue = enum {
    [0]Undefined,
    [1]Null,
    [2]String(String)
    [3]Bytes(Vec<u8>)
    [4]Other(String)
    [5]Exception(String)
}
pink::PinkEvent = enum {
    [2]SetHook {
        hook: pink::HookPoint,
        contract: ink_primitives::types::AccountId,
        selector: u32,
        gas_limit: u64,
    }
    [3]DeploySidevmTo {
        contract: ink_primitives::types::AccountId,
        code_hash: [u8; 32],
    }
    [4]SidevmMessage(Vec<u8>)
    [5]CacheOp(pink::CacheOp)
    [6]StopSidevm,
    [7]ForceStopSidevm {
        contract: ink_primitives::types::AccountId,
    }
    [8]SetLogHandler(ink_primitives::types::AccountId)
    [9]SetContractWeight {
        contract: ink_primitives::types::AccountId,
        weight: u32,
    }
    [10]UpgradeRuntimeTo {
        version: (u32, u32),
    }
    [11]SidevmOperation(pink::SidevmOperation)
    [12]SetJsRuntime([u8; 32])
}
pink::HookPoint = enum {
    [0]OnBlockEnd,
}
ink_primitives::types::AccountId = struct {
    : [u8; 32],
}
pink::CacheOp = enum {
    [0]Set {
        key: Vec<u8>,
        value: Vec<u8>,
    }
    [1]SetExpiration {
        key: Vec<u8>,
        expiration: u64,
    }
    [2]Remove {
        key: Vec<u8>,
    }
}
pink::SidevmOperation = enum {
    [0]Start {
        contract: ink_primitives::types::AccountId,
        code_hash: [u8; 32],
        workers: pink::Workers,
        config: pink::SidevmConfig,
    }
    [1]SetDeadline {
        contract: ink_primitives::types::AccountId,
        deadline: u32,
    }
}
pink::Workers = enum {
    [0]All,
    [1]List(Vec<[u8; 32]>)
}
pink::SidevmConfig = struct {
    max_code_size: u32,
    max_memory_pages: u32,
    vital_capacity: u64,
    deadline: u32,
}
pink::system::Error = enum {
    [0]PermisionDenied,
    [1]DriverNotFound,
    [2]CodeNotFound,
    [3]ConditionNotMet,
}
pink::system::DriverError = enum {
    [0]Other(String)
    [1]SystemError(pink::system::Error)
    [2]BadOrigin,
}
pink::system::CodeType = enum {
    [0]Ink,
    [1]Sidevm,
}
test_chain_extensions::Root = struct {
    _http_request: pink::chain_extension::http_request::HttpRequest,
    _http_response: pink::chain_extension::http_request::HttpResponse,
    _http_error: pink::chain_extension::http_request::HttpRequestError,
    _quota_error: pink::chain_extension::StorageQuotaExceeded,
    _error_code: pink::chain_extension::ErrorCode,
    _js_code: pink_types::js::JsCode,
    _js_value: pink_types::js::JsValue,
    _pink_event: pink::PinkEvent,
    _system_err: pink::system::Error,
    _driver_err: pink::system::DriverError,
    _code_type: pink::system::CodeType,
}

